libdnf/utils/string.cpp

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright (C) 2020-2021 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2.1 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#include "string.hpp"
+
+
+namespace libdnf::utils::string {
+
+
+std::vector<std::string> rsplit(const std::string & str, const std::string & delimiter, std::size_t limit) {
+    std::vector<std::string> result;
+
+    if (str.empty()) {
+        result.push_back("");
+        return result;
+    }
+
+    // let's make a copy and resize the string during the split - it's easier than messing with indexes
+    std::string str_copy(str);
+
+    std::size_t current = str_copy.rfind(delimiter);
+    const std::size_t delimiter_length = delimiter.size();
+    while (current != std::string::npos) {
+        if (limit != std::string::npos && result.size() + 1 >= limit) {
+            result.insert(result.begin(), str_copy);
+            return result;
+        }
+
+        result.insert(result.begin(), str_copy.substr(current + delimiter_length));
+        str_copy.resize(current);
+        current = str_copy.rfind(delimiter);
+    }
+
+    current = 0;
+    result.insert(result.begin(), str_copy);
+    return result;
+}
+
+
+std::vector<std::string> split(const std::string & str, const std::string & delimiter, std::size_t limit) {
+    std::vector<std::string> result;
+
+    std::size_t previous = 0;
+    std::size_t current = str.find(delimiter);
+    const std::size_t delimiter_length = delimiter.size();
+    while (current != std::string::npos) {
+        if (limit != std::string::npos && result.size() + 1 >= limit) {
+            result.push_back(str.substr(previous));
+            return result;
+        }
+
+        result.push_back(str.substr(previous, current - previous));
+        previous = current + delimiter_length;
+        current = str.find(delimiter, previous);
+    }
+    result.push_back(str.substr(previous, current - previous));
```

Comment:
`current` is always equal to `std::string::npos` on this line.
So using the term `current - previous` is not a good idea. The result is a big number, so it passes the tests.

Simply `result.push_back(str.substr(previous));`